# Image Processing & OpenCV 


## Outline 

1. Introduction
2. Warm up
3. OpenCV
    - How to install OpenCV
    - Basic operations
    - 

## Introduction: 

After taking a photo, we are so used to use some apps to add some photo effects or do some filter. In fact,a digital image is nothing more than a bunch of datas which indicating variations of red, green, and blue at a particular location on a grid of pixels. Most of the time, we just view this pixels as miniature rectangles and with little creative thinking on making some lower level maniupulation of pixels with code. In this tutorial, we will work on such pixels. Also, I will introduce `OpenCV` (Open Source Computer Vision),which is a library of programming functions mainly aimed at real-time computer vision.


## Warm up

In the first part, we will work with `PIL`, `numpy`, `matplotlib`, and have an overview on image data. 

In [1]:
# setup library imports
# import matplotlib.pyplot as plt
import numpy as np
import PIL
from PIL import Image

In [3]:
# load image from a image file
image = Image.open('buggy.jpg')
# change image pixel data to a three dimensional array 
image_array = np.asarray(image)

In [5]:
print image_array


Sample Output(shorten): 
```python      

        [[[  8  14  14]
  [  6  12  12]
  [  4  10   8]
  ..., 
  [108  19  13]
  [113  22  19]
  [117  26  25]]

 [[  9  13  12]
  [  6  10   9]
  [  5   9   8]
  ..., 
 [[ 94 117 125]
  [ 97 115 125]
  [100 117 127]
  ..., 
  [168 165 158]
  [168 164 155]
  [168 163 157]]]
```


The output is a  `three dimensional array `, which is an array within an array within an array. The outside dimension is the whole image, the second dimension correspond to each row, and the third dimension represent each pixel in that row. The inner list like ```python [ 73  71  30]  ``` is a `rgb` (red, green, blue) reprensentation. Some time the inner may like ```python [ 73  71  30 0]  ```,which is a `rgba` (red, green, blue,alpha) reprensentation.

In [5]:
#generate and show image from a three dimensional rbga array
plt.imshow(image_array)
plt.show()

Sample Output: 
<img src="buggy.jpg", height="200" width="400">


In [3]:
# print out width, height and total pixel number of the image
width, height = image.size
print "width: ", width
print "height: ", height
print "total pixel: ", width* height

width:  4454
height:  2505
total pixel:  11157270


Output: 
```python
width:  4454
height:  2505
total pixel:  11157270
```
So even though we can get a pixel array from an image, since the size of that array is big, the computational cost is really expensive when manipulating the pixel array directly. So we need some more powerful tool to deal with these pixel data and have some amazing output. `OpenCV` is a great tool in this situation, and in the following part, we will talk about `openCV - Pyhton`. 

## OpenCV

In [3]:
new_array =[[[0, 255, 0],[0, 255, 0],[0, 255, 0]],[[255, 0, 0],[0, 255, 0],[0, 255, 0]]]
plt.imshow(new_array)
plt.show()

### Install OpenCV-Python




- Installation in Mac  
     * Since we have already installed the Anaconda platform, we can open a new terminal window and type: 
```bash
        conda install -c https://conda.binstar.org/menpo opencv
```
   
  * If you don’t use Anaconda, then you can follow [https://www.youtube.com/watch?v=37RvqZVddAw](https://www.youtube.com/watch?v=37RvqZVddAw) to do the installation. 
  

 - Installation in Windows
 
     * PC user can follow this link if you have already installed the Anaconda:
     http://mathalope.co.uk/2015/05/07/opencv-python-how-to-install-opencv-python-package-to-anaconda-windows/
     
     * If not, then you can follow this link:
     http://docs.opencv.org/2.4/doc/tutorials/introduction/windows_install/windows_install.html
     
Unlike `matplotlib`, `numpy` and other packages we used before, it may take a while to figure out how to install openCV, but later you will find this tool is really powerful when deal with computer vision. 
    


         


### Overview 


`OpenCV-Python` is a library of Python bindings designed to solve computer vision problems.

`OpenCV-Python` makes use of `Numpy`, which is a highly optimized library for numerical operations with a `MATLAB-style syntax`. All the `OpenCV` array structures are converted to and from Numpy arrays. This also makes it easier to integrate with other libraries that use `Numpy` such as `SciPy` and `Matplotlib`.


### Basic Operations on Images

First, we will learn how to read,display, and save an image. 

- Read an image: `cv2.imread()`

- Show an image: `cv2.imshow() `

- Save an image: `cv2.imwrite() `

In [1]:
# setup library imports
import cv2
#Load buggy image
img_buggy = cv2.imread('buggy.jpg')
# your can try the below part by yourself
# TRY YOURSELF START
#Display an image in a window
# cv2.imshow('image',img_buggy)
#wait user to press any key
# cv2.waitKey(0)
#Destroys all the windows we created
# cv2.destroyAllWindows()

#TRY YOURSELF END

Then we will do some basic operations on the images. We can access a pixel value by its row and column coordinates.

**Importance notice: In openCV, pixel's representation is BGR (blue,green, red), instead of RGB(red,green,blue)**

Image processing usually follow the below procedure：
- load image in color mode:  `cv2.imread('image.jpg',cv2.IMREAD_COLOR)`
- convert image to grayscale mode:  `cv2.cvtColor(image,cv2.COLOR_BGR2GRAY) `
- doing image analysis
- apply the output back to the colored image and display

In the following part, we will begin to do some fancy things. 

In the next block, I want to put the CMU logo on the top-left corner of the buggy image. 

In [2]:
# Load two images
buggy_img = cv2.imread('buggy.jpg')
logo_img = cv2.imread('logo.jpg')

After loading two images, since we want to put the logo on the top-left corner of the buggy image, we need to know the logo size and get a image region in buggy image, then later we can replace the image region with logo.In openCV, we can only add two image with same size. 

In [4]:
# get the logo_image's size
rows,cols,channels = logo_img.shape
#then got a image region on the top-left corner of the buggy image
roi = buggy_img[0:rows, 0:cols ]

Then, we will convert the logo image from one color-space to another, in this case, from BGR to Gray. For BGR image, it returns an array of Blue, Green, Red values. For grayscale image, just corresponding intensity is returned. 

In OpenCV, there are three common color-space:  `BGR`, `GRAY` and `HSV`. For color conversion, we use the function `cv2.cvtColor(input_image, flag)` where `flag` determines the type of conversion.

For `BGR` -> `Gray` conversion we use the flags `cv2.COLOR_BGR2GRAY`. Similarly for `BGR` -> `HSV`, we use the flag `cv2.COLOR_BGR2HSV`. 

**Note: ** For HSV, Hue range is [0,179], Saturation range is [0,255] and Value range is [0,255]. Different softwares use different scales. So if you are comparing OpenCV values with them, you need to normalize these ranges.


_Why do we convert the color-space:_
- Grayscale image, just corresponding to the intensity. In our case, we want to cut out the logo part from the logo image, so we can only manipulate based on the intensity and it save memory and running time. 
- In HSV, it is more easier to represent a color than RGB color-space. So if we want to focus on the color part, we can covert to HSV space. 





In [5]:
# Now create a mask of logo and create its inverse mask also
img2gray = cv2.cvtColor(logo_img,cv2.COLOR_BGR2GRAY)


We often use threshold when dealing with grayscale image. Below is the pseudo code of a general thresholding. 

**Thresholding rule:**
```python
if src(x,y) > thresh
  dst(x,y) = maxValue
else
  dst(x,y) = 0
```

After we got the binary thresholding, we can applying binary thresholding to the input image. 

In [6]:
# create a binary thresholding
thresh = 125
maxValue = 255
ret, mask = cv2.threshold(img2gray, thresh, maxValue, cv2.THRESH_BINARY)
mask_inv = cv2.bitwise_not(mask)

# apply thresholding to buggy image and black-out the area of logo in ROI
img1_bg = cv2.bitwise_and(roi,roi,mask = mask)

# apply thresholding to logo image and take only region of logo from logo image.
img2_fg = cv2.bitwise_and(logo_img,logo_img,mask = mask_inv)

# Put logo in ROI and modify the buggy image
# OpenCV can only add two image with same size
# So make sure the images' size are match
dst = cv2.add(img1_bg,img2_fg)
#if is also works if you can wrote dst = img1_bg + img2_fg
buggy_img[0:rows, 0:cols ] = dst

# cv2.imshow('res',buggy_img)
# cv2.waitKey(0)
# cv2.destroyAllWindows()

Sample Input:

- Logo image
<img src="logo.jpg", height="200" width="400">
- Buggy image: 
<img src="buggy.jpg", height="200" width="400">

Sample Output: 
<img src="combined.png", height="200" width="400">

### Face Detection using Haar Cascades

#### Introduction to Cascade of Classifiers

"Instead of applying all the 6000 features on a window, group the features into different stages of classifiers and apply one-by-one. (Normally first few stages will contain very less number of features). If a window fails the first stage, discard it. We don't consider remaining features on it. If it passes, apply the second stage of features and continue the process. The window which passes all stages is a face region." - citation from [Face Detection using Haar Cascades](http://www.bogotobogo.com/python/OpenCV_Python/python_opencv3_Image_Object_Detection_Face_Detection_Haar_Cascade_Classifiers.php)


<img src="cascade.png", height="200" width="400">     <center> -image source from [How Face Detection Works](http://www.cognotics.com/opencv/servo_2007_series/part_2/sidebar.html)</center>

#### Using OpenCV's pre-trained classifier

In the following part, we will see the basics of face detection using Haar Feature-based Cascade Classifiers. OpenCV already contains many pre-trained classifiers for face, eyes, smile etc. These XML files are stored in `opencv/data/haarcascades/` folder. So, in the next part, we will first use the provided classifier and do some simply face detection. 

In [1]:
#Copy the 'haarcascade_frontalface_default.xml' into the current folder
# import the default face detection xml
face_cascade = cv2.CascadeClassifier('haarcascade_frontalface_default.xml')


In [ ]:
#import the buggy image
buggy_img = cv2.imread('buggy.jpg')
# create grayscale mode image 
gray = cv2.cvtColor(buggy_img, cv2.COLOR_BGR2GRAY)

"""
cv2.CascadeClassifier.detectMultiScale(image[, scaleFactor[, minNeighbors[, flags[, minSize[, maxSize]]]]]) 
- image : Matrix of the type CV_8U containing an image where objects are detected.
- scaleFactor : Parameter specifying how much the image size is reduced at each image scale. 
Suppose, the scale factor is 1.03, it means we're using a small step for resizing, i.e. reduce size by 3 %, 
we increase the chance of a matching size with the model for detection is found, while it's expensive.
- minNeighbors : Parameter specifying how many neighbors each candidate rectangle should have to retain it. 
This parameter will affect the quality of the detected faces: higher value results in less detections but with higher quality. 
"""

faces = face_cascade.detectMultiScale(gray, 1.2, 3)
#If faces are found, it returns the positions of detected faces as Rect(x,y,w,h).
for (x,y,w,h) in faces:
    cv2.rectangle(buggy_img,(x,y),(x+w,y+h),(255,0,0),2)
    roi_gray = gray[y:y+h, x:x+w]
    roi_color = buggy_img[y:y+h, x:x+w]
cv2.imshow('img',buggy_img)
cv2.waitKey(0)
cv2.destroyAllWindows()


Sample output: 
<img src="buggy_found.jpg", height="200" width="400">    

### Create our own Haar Cascade 

In the following part, we will make more fun and create our own Haar Cascade. We will construct a tartans's cascade classifier. 


[image-net.org](http://image-net.org/) is a great website to find source images. 
For example, if we want to find some sample image which contains human faces. 
type "face" in the searching bar -> select the wanted Synsets -> click "Downloads" -> click "URLs",then we will get a webpage contains bunch of images' urls. 

We can use urllib2 for python2 or urllib for python3 fetch images automatically from these urls. 

In [2]:
#import library
import urllib
import cv2
import numpy as np
import os 


In [3]:
# in this method, fetched some animal's images
# and stored the resized and grayscale images into "neg" file 
#and we will use these images as background images 
def store_raw_images():
    neg_image_link = 'http://image-net.org/api/text/imagenet.synset.geturls?wnid=n01316949'
    neg_image_urls=urllib.urlopen(neg_image_link).read().decode()
    if not os.path.exists('neg'):
        os.makedirs('neg')
    pic_num = 1
    for i in neg_image_urls.split('\n'):
        try:
            if pic_num <500:
                # print i can help us to see the progress 
                print i
                urllib.urlretrieve(i,"neg/"+str(pic_num)+".jpg")
                img = cv2.imread("neg/"+str(pic_num)+".jpg",cv2.IMREAD_GRAYSCALE)
                resized_img = cv2.resize(img,(100,100))
                cv2.imwrite("neg/"+str(pic_num)+".jpg",resized_img)
                pic_num +=1
            else: 
                break
        except Exception as e:
            print (str(e))  
    
store_raw_images()



http://www.horse-chiropractor.com/animal-chiropractor/equine_chiropractor.jpg
http://farm4.static.flickr.com/3142/2898261629_3af75a6fc0.jpg
http://farm4.static.flickr.com/3091/3162253667_1ddaf0a06f.jpg
http://www.bancoimagenes.com/cd738/cd738f155_a.jpg
/Users/jenkins/miniconda/0/2.7/conda-bld/work/opencv-2.4.11/modules/imgproc/src/imgwarp.cpp:1968: error: (-215) ssize.area() > 0 in function resize

http://static.andaluciaimagen.com/Burro-como-animal-de-transporte-51334.jpg
[Errno socket error] [Errno 8] nodename nor servname provided, or not known
http://farm1.static.flickr.com/30/62940360_e2bc08a608.jpg
http://www.engormix.com/images/s_articles/eficiencia03.jpg
http://static.flickr.com/41/124541184_97729d713d.jpg
http://perso.wanadoo.es/josanar620404/aAa.jpg
/Users/jenkins/miniconda/0/2.7/conda-bld/work/opencv-2.4.11/modules/imgproc/src/imgwarp.cpp:1968: error: (-215) ssize.area() > 0 in function resize

http://farm4.static.flickr.com/3224/3107362693_994b7545f7.jpg
http://farm1.static

OSError: [Errno 2] No such file or directory: 'uglies'

In [3]:
#find some more background images 
neg_images_link = 'http://image-net.org/api/text/imagenet.synset.geturls?wnid=n07942152'   
neg_image_urls=urllib.urlopen(neg_images_link).read().decode()
pic_num = 292

if not os.path.exists('neg'):
    os.makedirs('neg')

for i in neg_image_urls.split('\n'):
    try:
        if pic_num <=700:
            print(i)

            urllib.urlretrieve(i, "neg/"+str(pic_num)+".jpg")
            img = cv2.imread("neg/"+str(pic_num)+".jpg",cv2.IMREAD_GRAYSCALE)
            # should be larger than samples / pos pic (so we can place our image on it)
            resized_image = cv2.resize(img, (100, 100))
            cv2.imwrite("neg/"+str(pic_num)+".jpg",resized_image)
            pic_num += 1
        else:
            break

    except Exception as e:
        print(str(e)) 

http://farm4.static.flickr.com/3186/2606823711_b07495378f.jpg
http://farm4.static.flickr.com/3360/3236502886_6a505502c5.jpg
http://farm5.static.flickr.com/4022/4322083945_bc2db8ccf3.jpg
http://farm4.static.flickr.com/3121/2556445732_0f905b11e4.jpg
http://farm4.static.flickr.com/3551/3443533285_2a70b92f2c.jpg
http://farm1.static.flickr.com/224/491407069_8dbdc006c4.jpg
http://farm1.static.flickr.com/33/42539714_c899059d27.jpg
http://farm4.static.flickr.com/3512/3248952529_0bda5b70eb.jpg
http://farm3.static.flickr.com/2461/3707202944_74e7a04f77.jpg
http://farm4.static.flickr.com/3053/2664253302_956cb8356d.jpg
/Users/jenkins/miniconda/0/2.7/conda-bld/work/opencv-2.4.11/modules/imgproc/src/imgwarp.cpp:1968: error: (-215) ssize.area() > 0 in function resize

http://farm4.static.flickr.com/3236/3060322139_39a255d95c.jpg
http://farm1.static.flickr.com/150/406427892_3776e7e6ef.jpg
http://farm4.static.flickr.com/3609/4008974057_d0d7989bda.jpg
http://farm4.static.flickr.com/3055/2658058589_618ac5

Then we will find that there are some white picture and have some text on it. This is because these images are no longer available, instead of serving and HTTP error. We can create a file called 'uglies' and drag one of the "dirty" into this file. Then, we call this find_uglies() method to find all instances of this "dirty" images and remove them.  

In [4]:
#delete "dirty" images      
def find_uglies():
    match = False
    for file_type in ['neg']:
        for img in os.listdir(file_type):
            for ugly in os.listdir('uglies'):
                try:
                    current_image_path = str(file_type)+'/'+str(img)
                    ugly = cv2.imread('uglies/'+str(ugly))
                    question = cv2.imread(current_image_path)
                    if ugly.shape == question.shape and not(np.bitwise_xor(ugly,question).any()):
                        print('That is one ugly pic! Deleting!')
                        print(current_image_path)
                        os.remove(current_image_path)
                except Exception as e:
                    print(str(e))
find_uglies()

'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
That is one ugly pic! Deleting!
neg/295.jpg
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
That is one ugly pic! Deleting!
neg/296.jpg
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
'NoneType' object has no attribute 'shape'
That is one ugly pic! Deleting!
neg/297.jpg
'NoneType' object has no attribute 'shape'
'NoneTyp

Then we need to create a description txt "bg.txt" for these negative images.

In [5]:
def create_pos_n_neg():
    for file_type in ['neg']: 
        for img in os.listdir(file_type):
            if file_type == 'neg':
                line = file_type+'/'+img+'\n'
                with open('bg.txt','a') as f:
                    f.write(line)
create_pos_n_neg()

In [ ]:
#create sample image 
apple_img = cv2.imread('apple.jpg')
gray_apple = cv2.cvtColor(apple_img, cv2.COLOR_BGR2GRAY)
resized_img = cv2.resize(gray_apple,(50,50))
cv2.imwrite("sample_apple.jpg",resized_img)

Here, I used an apple's image and later will detect apples in test image. My experience is crop the image and try best to let the target object fill more space in the whole image. By doing so, we can have a more accurate result. 
<img src="apple.png", height="100" width="100"> 


In the following part, we will works on terminal. 
- First, in the terminal window, type "git clone https://github.com/Itseez/opencv.git" to fetch openCV. 
- Second, copy `"createsamples.cpp"` into current working space. We can find `"createsamples.cpp"` in `"/opencv-2.4.13 2/apps/haartraining"`. 
- Third, type `"mkdir info"`. This is the place where we will stuff all of the positive images.
- Fourth, type `"mkdir data"`. This is the place where we will store all the training output. 


Now, we have done all the preparation works. Now, go ahead and begin training the samples. 
- In the terminal, type `"opencv_createsamples -img sample_apple.jpg -bg bg.txt -info info/info.lst -pngoutput info -maxxangle 0.5 -maxyangle 0.5 -maxzangle 0.5 -num 360 -w 30 -h 30 -vec positives.vec"`
- Using this You can find command line arguments' meaning here: [http://docs.opencv.org/3.1.0/dc/d88/tutorial_traincascade.html](http://docs.opencv.org/3.1.0/dc/d88/tutorial_traincascade.html)


Sample output: 
<img src="terminal_1.png", height="500" width="800"> 

So far,we have already create sample images and a positive vector. Now,we can start the training process. To do the training, in the terminal, type  `"opencv_traincascade -data data -vec positives.vec -bg bg.txt -numPos 300 -numNeg 150 -numStages 10 -w 30 -h 30"`

Sample output: 
<img src="terminal_2.png", height="500" width="800"> 

Then, in our "data" file, we can find `cascade.xml`. That is the our own "Haar Cascade classifier". Now, we can go ahead to the detection stage.  

In [ ]:
# import our own cascade.xml, here I renamed it to "apple_cascade.xml"
apple_cascade=cv2.CascadeClassifier('apple_cascade.xml')

apple_img = cv2.imread('apple_target1.jpg')
gray = cv2.cvtColor(apple_img, cv2.COLOR_BGR2GRAY)
apple = apple_cascade.detectMultiScale(gray, 1.1,3)
for (x,y,w,h) in apple:
    cv2.rectangle(apple_img,(x,y),(x+w,y+h),(255,255,0),2)
    roi_gray = gray[y:y+h, x:x+w]
    roi_color = apple_img[y:y+h, x:x+w]
cv2.imshow('img',apple_img)
cv2.waitKey(0)
cv2.destroyAllWindows()


Sample output: 
<img src="output.png", height="100" width="200"> 

We can see the output is not that accurate. We can increase the accuracy by adding more background image, or more positive images. Also, we can change the command line arguments's value. For example, we use "-w 30 -h 30", we can increase these numbers and it will be more expensive. Also, we can change "-numPos 300 -numNeg 150". 

## Useful Links

Resource for Haar Cascades: 
- http://www.bogotobogo.com/python/OpenCV_Python/python_opencv3_Image_Object_Detection_Face_Detection_Haar_Cascade_Classifiers.php
- http://www.cognotics.com/opencv/servo_2007_series/part_2/sidebar.html
- http://docs.opencv.org/3.1.0/dc/d88/tutorial_traincascade.html
- https://pythonprogramming.net/haar-cascade-object-detection-python-opencv-tutorial/

Resource for openCV library 
- [http://docs.opencv.org/3.0-beta/doc/py_tutorials/py_tutorials.html](http://docs.opencv.org/3.0-beta/doc/py_tutorials/py_tutorials.html)

#### Hope you enjoy this tutorial. Cheers!